integration/test/Test/Connection.hs

```diff
@@ -0,0 +1,403 @@
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+module Test.Connection where
+
+import API.Brig (getConnection, postConnection, putConnection)
+import API.BrigInternal
+import API.Galley
+import SetupHelpers
+import Testlib.Prelude
+import UnliftIO.Async (forConcurrently_)
+
+testConnectWithRemoteUser :: HasCallStack => Domain -> App ()
+testConnectWithRemoteUser owningDomain = do
+  (alice, bob, one2oneId) <- createOne2OneConversation owningDomain
+  aliceId <- alice %. "qualified_id"
+  getConversation alice one2oneId `bindResponse` \resp -> do
+    resp.status `shouldMatchInt` 200
+    others <- resp.json %. "members.others" & asList
+    qIds <- for others (%. "qualified_id")
+    qIds `shouldMatchSet` ([] :: [Value])
+  void $ putConnection bob alice "accepted" >>= getBody 200
+  getConversation bob one2oneId `bindResponse` \resp -> do
+    resp.status `shouldMatchInt` 200
+    others <- resp.json %. "members.others" & asList
+    qIds <- for others (%. "qualified_id")
+    qIds `shouldMatchSet` [aliceId]
+
+testRemoteUserGetsDeleted :: HasCallStack => App ()
+testRemoteUserGetsDeleted = do
+  alice <- randomUser OwnDomain def
+
+  charlieConnected <- do
+    charlie <- randomUser OtherDomain def
+    connectTwoUsers alice charlie
+    pure charlie
+
+  charliePending <- do
+    charlie <- randomUser OtherDomain def
+    -- the connection should be pending here
+    postConnection alice charlie `bindResponse` \resp ->
+      resp.status `shouldMatchInt` 201
+
+    getConnection alice charlie `bindResponse` \resp -> do
+      resp.json %. "status" `shouldMatch` "sent"
+      resp.status `shouldMatchInt` 200
+
+    getConnection charlie alice `waitForResponse` \resp -> do
+      resp.json %. "status" `shouldMatch` "pending"
+      resp.status `shouldMatchInt` 200
+
+    pure charlie
+
+  charlieBlocked <- do
+    charlie <- randomUser OtherDomain def
+    postConnection alice charlie `bindResponse` \resp ->
+      resp.status `shouldMatchInt` 201
+
+    putConnection charlie alice "blocked" `bindResponse` \resp ->
+      resp.status `shouldMatchInt` 200
+
+    getConnection charlie alice `bindResponse` \resp -> do
+      resp.json %. "status" `shouldMatch` "blocked"
+      resp.status `shouldMatchInt` 200
+
+    pure charlie
+
+  charlieUnconnected <- do
+    randomUser OtherDomain def
+
+  forConcurrently_ [charliePending, charlieConnected, charlieBlocked, charlieUnconnected] \charlie -> do
+    deleteUser charlie
+
+    -- charlie is on their local backend, so asking should be instant
+    getConnection charlie alice `bindResponse` \resp ->
+      resp.status `shouldMatchInt` 404
+
+    -- for alice, charlie is on the remote backend, so the status change
+    -- may not be instant
+    getConnection alice charlie `waitForResponse` \resp ->
+      resp.status `shouldMatchInt` 404
```

Comment:
I suppose this works too, but we used a different approach so far. In particular, Alice would wait for a notification on her web socket that Charlie got deleted. Once she has that, she is sure that `getConnection alice charlie` returns a 404.

Reply:
does this work if the connection is pending?

Reply:
I don't think it does as there are no notifications for this state.

I'd say it's fine to do it this way; I was just saying so far we have done it differently, but that's no argument against doing something new like you did here.